notebooks/source/bayesian_hierarchical_linear_regression.ipynb

```diff
@@ -0,0 +1,677 @@
+    "\n",
+    "Here, I'll use the middle ground: **Partial pooling**. Specifically, I'll assume that while $\\alpha$'s and $\\beta$'s are different for each patient as in the unpooled case, **the coefficients all share similarity**. We can model this by assuming that each individual coefficient comes from a common group distribution. The image below represents this model graphically:\n",
+    "\n",
+    "<img src=\"https://i.ibb.co/H7NgBfR/Artboard-2-2x-100.jpg\" alt=\"drawing\" width=\"600\"/>\n",
+    "\n",
+    "Mathematically, the model is described by the following equations:\n",
+    "$$\n",
+    "\\mu_{\\alpha} \\sim \\mathcal{N}(0, 100) \\\\\n",
+    "\\sigma_{\\alpha} \\sim |\\mathcal{N}(0, 100)| \\\\\n",
+    "\\mu_{\\beta} \\sim \\mathcal{N}(0, 100) \\\\\n",
+    "\\sigma_{\\beta} \\sim |\\mathcal{N}(0, 100)| \\\\\n",
+    "\\alpha_i \\sim \\mathcal{N}(\\mu_{\\alpha}, \\sigma_{\\alpha}) \\\\\n",
+    "\\beta_i \\sim \\mathcal{N}(\\mu_{\\beta}, \\sigma_{\\beta}) \\\\\n",
+    "\\sigma \\sim \\mathcal{N}(0, 100) \\\\\n",
+    "FVC_{ij} \\sim \\mathcal{N}(\\alpha_i + t \\beta_i, \\sigma)\n",
+    "$$\n",
+    "\n",
+    "where *t* is the time in weeks. Those are very uninformative priors, but that's ok: our model will converge!\n",
+    "\n",
+    "Implementing this model in Numpyro is pretty straightforward:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import numpyro\n",
+    "from numpyro.infer import MCMC, NUTS, Predictive\n",
+    "import numpyro.distributions as dist\n",
+    "from jax import random"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def model(PatientID, Weeks, FVC_obs=None):\n",
+    "    mu_a = numpyro.sample(\"mu_a\", dist.Normal(0., 100.))\n",
+    "    sigma_a = numpyro.sample(\"sigma_a\", dist.HalfNormal(100.))\n",
+    "    mu_b = numpyro.sample(\"mu_b\", dist.Normal(0., 100.))\n",
+    "    sigma_b = numpyro.sample(\"sigma_b\", dist.HalfNormal(100.))\n",
+    "    \n",
+    "    unique_patient_IDs = np.unique(PatientID)\n",
+    "    n_patients = len(unique_patient_IDs)\n",
+    "    \n",
+    "    a = numpyro.sample(\"a\", dist.Normal(mu_a, sigma_a).expand([n_patients]))\n",
+    "    b = numpyro.sample(\"b\", dist.Normal(mu_b, sigma_b).expand([n_patients]))\n",
+    "    \n",
+    "    sigma = numpyro.sample(\"sigma\", dist.HalfNormal(100.))\n",
+    "    \n",
+    "    FVC_est = a[PatientID] + b[PatientID] * Weeks\n",
+    "    \n",
+    "    with numpyro.plate(\"data\", len(PatientID)):\n",
+    "        numpyro.sample(\"obs\", dist.Normal(FVC_est, sigma), obs=FVC_obs)"
```

Comment:
The nice thing about Pyro is its corresponding 1-1 to a plate diagram. Because you did draw the diagram, it would be nice to reorder the model to something like
```python
    mu_a = numpyro.sample("mu_a", dist.Normal(0., 100.))
    sigma_a = numpyro.sample("sigma_a", dist.HalfNormal(100.))
    mu_b = numpyro.sample("mu_b", dist.Normal(0., 100.))
    sigma_b = numpyro.sample("sigma_b", dist.HalfNormal(100.))
    sigma = numpyro.sample("sigma", dist.HalfNormal(100.))

    unique_patient_IDs = np.unique(PatientID)
    n_patients = len(unique_patient_IDs)
    with numpyro.plate("plate_i", n_patients):
        a = numpyro.sample("a", dist.Normal(mu_a, sigma_a))
        b = numpyro.sample("b", dist.Normal(mu_b, sigma_b))
        with numpyro.plate("plate_j", len(PatientID)):
            numpyro.sample("obs", dist.Normal(FVC_est, sigma), obs=FVC_obs)
```

Btw, should it be FVC_ji rather than FVC_ij in the plot and the formula? where j is data dimension and i is patient id dimension.

Reply:
Hi! Thanks! I implemented all suggestions, except this one. I couldn't make it work. My attempt was this:
```Python
def model(PatientID, Weeks, FVC_obs=None):
    mu_a = numpyro.sample("mu_a", dist.Normal(0., 100.))
    sigma_a = numpyro.sample("sigma_a", dist.HalfNormal(100.))
    mu_b = numpyro.sample("mu_b", dist.Normal(0., 100.))
    sigma_b = numpyro.sample("sigma_b", dist.HalfNormal(100.))
    sigma = numpyro.sample("sigma", dist.HalfNormal(100.))

    unique_patient_IDs = np.unique(PatientID)
    n_patients = len(unique_patient_IDs)
    for i in numpyro.plate("plate_i", n_patients):
        a = numpyro.sample(f"a_{i}", dist.Normal(mu_a, sigma_a))
        b = numpyro.sample(f"b_{i}", dist.Normal(mu_b, sigma_b))
        FVC_est = a + b * Weeks[PatientID == i]
        
        with numpyro.plate("plate_j", (PatientID == i).sum()):
            numpyro.sample("obs", dist.Normal(FVC_est, sigma), obs=FVC_obs[PatientID == i])
```

but NumPyro is different than Pyro in non-vectorized plate (doesn't have the `for` loop). Without that, I don't know how to compute `FVC_est` for each patient at a time. How would that be possible?
Thanks!

Reply:
Oops... I missed your code `FVC_est = a[PatientID] + b[PatientID] * Weeks` before `plate_j`. Does it work for you?

Reply:
It doesn't, generates an error. The code
```Python
def model(PatientID, Weeks, FVC_obs=None):
    mu_a = numpyro.sample("mu_a", dist.Normal(0., 100.))
    sigma_a = numpyro.sample("sigma_a", dist.HalfNormal(100.))
    mu_b = numpyro.sample("mu_b", dist.Normal(0., 100.))
    sigma_b = numpyro.sample("sigma_b", dist.HalfNormal(100.))
    sigma = numpyro.sample("sigma", dist.HalfNormal(100.))

    unique_patient_IDs = np.unique(PatientID)
    n_patients = len(unique_patient_IDs)
    with numpyro.plate("plate_i", n_patients):
        a = numpyro.sample("a", dist.Normal(mu_a, sigma_a))  # (176,)
        b = numpyro.sample("b", dist.Normal(mu_b, sigma_b))  # (176,)
        FVC_est = a[PatientID] + b[PatientID] * Weeks  # (1549,)
        with numpyro.plate("plate_j", len(PatientID)):
            numpyro.sample("obs", dist.Normal(FVC_est, sigma), obs=FVC_obs)
```
The last line breaks with **ValueError: Incompatible shapes for broadcasting: ((1549, 176), (1549, 1549))**. I don't know how to fix. I understand what's going on with the first plate, but I still don't understand what happens in the background with the 2nd nested plate...

Reply:
If I "unnest" the 2nd plate, it works:
```Python
def model(PatientID, Weeks, FVC_obs=None):
    mu_a = numpyro.sample("mu_a", dist.Normal(0., 100.))
    sigma_a = numpyro.sample("sigma_a", dist.HalfNormal(100.))
    mu_b = numpyro.sample("mu_b", dist.Normal(0., 100.))
    sigma_b = numpyro.sample("sigma_b", dist.HalfNormal(100.))
    sigma = numpyro.sample("sigma", dist.HalfNormal(100.))

    unique_patient_IDs = np.unique(PatientID)
    n_patients = len(unique_patient_IDs)
    
    with numpyro.plate("plate_i", n_patients):
        a = numpyro.sample("a", dist.Normal(mu_a, sigma_a))
        b = numpyro.sample("b", dist.Normal(mu_b, sigma_b))
        FVC_est = a[PatientID] + b[PatientID] * Weeks
    
    with numpyro.plate("plate_j", len(PatientID)):
        numpyro.sample("obs", dist.Normal(FVC_est, sigma), obs=FVC_obs)
```
This works! Should I submit with this or try to make it work with nested plates? Cheers!

Reply:
If I understood the logic behind the nested plates correctly, they would only work if I'd reshape the data from a 2D table, where each row contains a tuple PatientID, Week, and FVC, into a 3D tensor, where each dimension would represent one thing (PatientIDs, Weeks, and FVCs). But making such a reshape would not be simple, as different patients have FVC measurements done in different weeks (I'd have to appropriately pad the 3D tensor to make it work)... does that make sense?

Reply:
I think I understand what is going on now. The FVC is partially observed so you flatten dimensions `i x j`. To use 2 `plate` statements as the above, you will need to reshape your data `FVC_obs` into `num_weeks x num_patients` with `NaN` for missing entries and mask out the likelihood `dist.Normal(FVC_est, sigma).mask(np.isfinite(FVC_obs))`. Then the model corresponds to the plate diagram. But it is expensive (in NumPyro, for masked distribution, we compute the full log probabilities, then mask out the result) because `FVC_obs` is sparse. Your original model is better so please keep it as-is. Sorry for making confusion!

nit: you can replace
```python
a = numpyro.sample("a", dist.Normal(mu_a, sigma_a).expand([n_patients]))
b = numpyro.sample("b", dist.Normal(mu_b, sigma_b).expand([n_patients]))
```
by
```python
    with numpyro.plate("plate_i", n_patients):
        a = numpyro.sample("a", dist.Normal(mu_a, sigma_a))
        b = numpyro.sample("b", dist.Normal(mu_b, sigma_b))
```
which is more *Pyronic*. :D

Reply:
Yes, that makes sense to me. Thanks for explanation!

Reply:
Done, thx! Btw, I replaced the variables with their correspondent greek letters, I think it's more readable that way... pls let me know what you think! Cheers!

Reply:
> it's more readable

Yeah, I agree.

Just the last nit: could you revert `plate_j` back to `data`? It is not `plate_j` from the diagram, which has length `t` (W1, W2,..., Wt). I think it is a subset of the "flatten" version of the joined `plate_j x plate_i`. Sorry for the language, but I guess you know what I mean. What do you think? Do I interpret it correctly?